Sources/SwiftLexicalLookup/SequentialScopeSyntax.swift

```diff
@@ -0,0 +1,93 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftSyntax
+
+/// Scope that, in addition to names introduced by itself,
+/// also handles names introduced by
+/// `IntroducingToSequentialParentScopeSyntax` children scopes.
+protocol SequentialScopeSyntax: ScopeSyntax {
+  /// Returns names introduced by `codeBlockItems`
+  /// and included `IntroducingToSequentialParentScopeSyntax` children
+  /// scopes that match the lookup.
+  func sequentialLookup(
+    in codeBlockItems: any Collection<CodeBlockItemSyntax>,
+    for identifier: Identifier?,
+    at syntax: SyntaxProtocol,
+    with config: LookupConfig,
+    state: LookupState,
+    createResultsForThisScopeWith getResults: ([LookupName]) -> (LookupResult)
+  ) -> [LookupResult]
+}
+
+extension SequentialScopeSyntax {
+  func sequentialLookup(
+    in codeBlockItems: any Collection<CodeBlockItemSyntax>,
+    for identifier: Identifier?,
+    at syntax: SyntaxProtocol,
+    with config: LookupConfig,
+    state: LookupState,
+    createResultsForThisScopeWith getResults: ([LookupName]) -> (LookupResult)
+  ) -> [LookupResult] {
+    var result = [LookupResult]()
+    var currentChunk = [LookupName]()
+
+    for codeBlockItem in codeBlockItems {
```

Comment:
`skipSequentialIntroductionFrom` was only [set when triggering lookup from within `guard ... else` clause](https://github.com/swiftlang/swift-syntax/blob/059860a66135ee92954ab42c34d82a7100a88151/Sources/SwiftLexicalLookup/ScopeImplementations.swift#L342). The condition would still hold for lookup that started there.

But I liked the idea of removing `skipSequentialIntroductionFrom` as it added another layer of complexity to sequential lookup, so I moved the check for lookup triggered in `else` body to [`introducesToSequentialParent` in `guard` scope implementation](https://github.com/swiftlang/swift-syntax/blob/252d3be80c010add2bac9f9712ac32d9892ea574/Sources/SwiftLexicalLookup/ScopeImplementations.swift#L321-L322). 

Should we now remove `LookupState`, or have it available across the API in case we need it in the future?

Reply:
I was also thinking about performance here and that code block items after the lookup position can’t be relevant but I forgot that there could be a local class declaration after the lookup position that the lookup position refers to.

Reply:
That's why I added now the first loop to scan for class declarations (as we need them in the first code block result to be available in the whole scope no matter the partition). Because to this change, we can now break early. Thanks for the suggestion!